### PR TITLE
job_controller: refactor job controller to be able to inject FakeClock for UTs

### LINKED ISSUE #110697: Flaky test TestSyncPastDeadlineJobFinished
### Which jobs are flaking?

pull-kubernetes-unit

### Which tests are flaking?

TestSyncPastDeadlineJobFinished

### Since when has it been flaking?

June 14th

### Testgrid link

https://storage.googleapis.com/k8s-triage/index.html?pr=1&text=TestSyncPastDeadlineJobFinished&job=pull-kubernetes-unit

### Reason for failure (if possible)

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/110638/pull-kubernetes-unit/1539289648939929600

```
=== FAIL: pkg/controller/job TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set (3.00s)
I0621 17:06:06.298308   54038 job_controller.go:190] Starting job controller
I0621 17:06:06.298379   54038 shared_informer.go:255] Waiting for caches to sync for job
I0621 17:06:06.298492   54038 shared_informer.go:262] Caches are synced for job
I0621 17:06:06.352546   54038 job_controller.go:504] enqueueing job default/job1
E0621 17:06:06.353048   54038 tracking_utils.go:109] "deleting tracking annotation UID expectations" err="couldn't create key for object default/job1: could not find key for obj \"default/job1\"" job="default/job1"
    job_controller_test.go:1892: Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in {[] <nil> <nil> 0 0 0  nil <nil>}
    --- FAIL: TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set (3.00s)
```

### Anything else we need to know?

Probably after #110294

### Relevant SIG(s)

/sig apps

### PR DESCRIPTION
**Pull request by harshanarayana**
#### What type of PR is this?
/kind cleanup
/kind flake
/kind failing-test
/kind regression

#### What this PR does / why we need it:
This PR takes inspiration from the `daemon_controller.go` on the KCM and refactors the `job_controller.go` to accept a `clock.Clock` as the argument in order to be able to provide a `testingclock.FakeClock` as the value during the Unit Test
and be able to simulate the cases of timing requirements. 

This also modifies the `job_controller_test.go` in order to be able to inject the `FakeClock`. This will help us get a better test behavior for items getting queued back into the `workqueue`

Currently, this only refactors the test `TestSyncPastDeadlineJobFinished`. Rest of the tests still use the default `clock.RealClock{}` while setting up the controller to no change the current behavior in place. 

#### Which issue(s) this PR fixes:

Fixes #110697 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

### CODE REVIEW

**Comment by harshanarayana** [comment]
>  Is this still a draft or not?

@alculquicondor Opening it up for review. Thanks.

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:821]: instead, make `newCondition` a method of the `Controller` struct

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:58]:
```suggestion
	clocktesting "k8s.io/utils/clock/testing"
```

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:114]: Instead of copying this implementation, it's better to add a function to provide a queue in `k8s.io/client-go/util/workqueue`. Then `NewController` can directly pass the clock interface too.

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:1281]: It actually makes more sense to pass `now` here than the clock.

File: pkg/controller/job/job_controller.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:821]: /Ack

Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:821]: /Ack

Doing this also means I will have to pull `ensureJobConditionStatus` into the `jm` as well. Or I have to keep passing the clock around still. Will do that as well.

File: pkg/controller/job/job_controller.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:833]: Moved `ensureJobConditionStatus` to the `jm` instead. This way, I can avoid the additional arg.

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:58]: /Ack

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:114]:
```golang
func NewNamedRateLimitingQueueWithCustomClock(clock clock.WithTicker, rateLimiter RateLimiter, name string) RateLimitingInterface {
	return &rateLimitingType{
		DelayingInterface: NewDelayingQueueWithCustomClock(clock, name),
		rateLimiter:       rateLimiter,
	}
}
```
Will add something like so?

File: pkg/controller/job/job_controller.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:1281]: This and `ensureJobConditionStatus` are a bit of a problematic one to refactor. They need to go hand in hand to be able to move them into the `jm`. Which one would you prefer? Sending `now` instead of the clock to both or moving them both into `jm` together ?

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:139]: /Ack

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1914]: /Ack

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:821]: Oops, this was an old idea. Better pass the clock instead.

Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:821]: Done

File: pkg/controller/job/job_controller.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:833]: /Done. Took care of it via https://github.com/kubernetes/kubernetes/pull/110710#discussion_r906291240

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go]: Done

File: pkg/controller/job/job_controller.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:1281]: DOne

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go]: Done

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1886]: Done

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go]: Done

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:116]:
```suggestion
	return newControllerFromClientWithClock(kubeClient, resyncPeriod, realClock)
```

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [reply pkg/controller/job/job_controller_test.go:1887]: I don't think this channel is needed

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:1902]:
```suggestion
			manager.clock.Sleep(time.Second)
```

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1887]: Done

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:116]: Done

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1890]: Done.

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1902]: Done

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go]:
```bash
=== Failed
=== FAIL: pkg/controller/job TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set (3.00s)
I0625 07:21:21.640508   53938 job_controller.go:194] Starting job controller
I0625 07:21:21.640672   53938 job_controller.go:508] enqueueing job default/job1
I0625 07:21:21.640775   53938 shared_informer.go:255] Waiting for caches to sync for job
I0625 07:21:21.641109   53938 shared_informer.go:262] Caches are synced for job
I0625 07:21:21.643565   53938 job_controller.go:508] enqueueing job default/job1
I0625 07:21:21.643740   53938 job_controller.go:508] enqueueing job default/job1
    job_controller_test.go:1911: Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in {[] 0001-01-01 00:00:01 +0000 UTC <nil> 0 0 0  nil 0xc000e0c3bc}
    --- FAIL: TestSyncPastDeadlineJobFinished/New_job_created_without_start_time_being_set (3.00s)

=== FAIL: pkg/controller/job TestSyncPastDeadlineJobFinished (3.31s)
```

Test seemingly started failing again after I did this refactoring,

**Review by alculquicondor — COMMENTED** [review]
/lgtm

File: pkg/controller/job/job_controller.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:1240]: Done

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go]: Not sure. Let me retry it again a few times to confirm. Loop test locally did not reproduce the issue though

**Comment by harshanarayana** [comment]
@alculquicondor Updated the changes suggested above. Please add an LGTM again when time permits.

**Comment by harshanarayana** [comment]
rebased to account for the changes that went into https://github.com/kubernetes/kubernetes/pull/111475

cc @alculquicondor @soltysh PTAL when time permits.

**Comment by alculquicondor** [comment]
> /kind failing-test

Do you have proof that this test was failing?

**Comment by harshanarayana** [comment]
> Do you have proof that this test was failing after https://github.com/kubernetes/kubernetes/issues/110697 was merged?

@alculquicondor The flaky tests were partially because of another issue as well where the wait for cache sync was not being performed. Which you addressed in one of your PR. But the changes done in this PR also addresses the original comment from a different PR around the `job_controller.go`

https://github.com/kubernetes/kubernetes/pull/110294#issuecomment-1148802029

Do you think these changes are no more required ?

**Comment by alculquicondor** [comment]
I'm ok with this changes, but it's more appropriate to remove these:

/remove-kind regression
/remove-kind flaky-test

This is just a cleanup

/priority backlog

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:439]:
```suggestion
			now := jm.clock.Now()
			passed := now.Sub(curJob.Status.StartTime.Time)
```

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:1868]: nit:
```suggestion
	fakeClock := clocktesting.NewFakeClock(time.Now())
```

Otherwise you might be starting with a unix timestamp of 0. Then I'm not sure what is happening below when you say `Unix()-1`

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:1899]:
```suggestion
			job.Spec.ActiveDeadlineSeconds = pointer.Int64(1)
```

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:1915]: let's use `PollImmediate`

File: pkg/controller/job/job_controller.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:439]: /Ack. Changes done accordingly.

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1868]: /Ack. Changed accordingly.

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1899]: /Ack.

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1915]: /Ack

**Comment by harshanarayana** [comment]
@alculquicondor PTAL when time permits. The changes have been addressed.

**Review by alculquicondor — COMMENTED** [review]
/lgtm

But this can wait until 1.26

**Comment by harshanarayana** [comment]
> /lgtm
> 
> But this can wait until 1.26

@alculquicondor Do you think we can take this forward in the 1.26 or should we wait a little longer ?

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:1923]: We could check at this point that there is no deadline condition here

**Review by alculquicondor — COMMENTED** [review]
Final nit

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1923]: @alculquicondor Sorry about the delay. I totally missed this comment/notification from you. Just pushed an updated version of the changes as suggested by you. PTAL when possible.

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:444]:
```suggestion
			passed := jm.clock.Since(curJob.Status.StartTime.Time)
```

**Review by alculquicondor — COMMENTED** [review]
minor suggestion
/approve

File: pkg/controller/job/job_controller.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller.go]: Done. PTAL When possible. cc @alculquicondor

**Review by alculquicondor — COMMENTED** [review]
/lgtm

File: pkg/controller/job/job_controller.go
Comment (soltysh) [inline pkg/controller/job/job_controller.go:133]: For the purpose of changing this clock only in unit tests I'd NOT expose this here, but only inside the `Controller` struct, to which unit tests have access since that's how we modify other bits, like podController or alike.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:133]: not sure if it can be done without explicitly recreating the queues. Is that fine?

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:133]: Since there is this open question and this is not urgent, we can leave it for after code freeze.

@harshanarayana can you make sure put this change through after code freeze?

Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:133]: @alculquicondor Sure. I can do that. Do you want me to create a different Issue to track this or just use the current one itself for all of it? Since we have no conclusion on this yet, I am open to either of the options. Or hold this changes from getting in until we conclude and address the required changes that @soltysh has in mind.

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:133]: just the current one. Let's wait for Maciej's answer.

Comment (mimowo) [reply pkg/controller/job/job_controller.go:133]: I remember solving a similar issue by introduction of the interim function `NewPodGCInternal` used to call from tests with non-default values for `gcCheckPeriod` and `quarantineTime`: https://github.com/kubernetes/kubernetes/blob/96fd7b1b50362858a621c03a16e21b71f64cfcbf/pkg/controller/podgc/gc_controller.go#L82

**Comment by soltysh** [comment]
/lgtm cancel
@harshanarayana the comment about not exposing clock as a parameter to `NewJobController` still holds, please ping me on slack when you get it fixed

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by harshanarayana** [comment]
@dims  This is still required. I got busy with release work and havent been able to get this one done. Will be taking care of this and pushing changes with rebased content.

**Comment by mimowo** [comment]
Having the fake clock will be useful for testing backoff behavior: https://github.com/kubernetes/kubernetes/pull/114516.. Some changes (as the one discussed in this thread: https://github.com/kubernetes/kubernetes/pull/114516#discussion_r1052790539) now require manual testing on a live cluster.

**Comment by harshanarayana** [comment]
@soltysh  @alculquicondor @mimowo Pushed an updated set of the changes based on the suggestion from @soltysh  and @mimowo PTAL when time permits.

File: pkg/controller/job/job_controller.go
Comment (mimowo) [inline pkg/controller/job/job_controller.go:1296]: nit: Please move this function where it was (below `pastBackoffLimitOnFailure`)

File: pkg/controller/job/job_controller_test.go
Comment (mimowo) [inline pkg/controller/job/job_controller_test.go:1883]: It seems `err != nil` would be unexpected here , maybe returning it `return false, err` to leave trace of the error if it happens.

**Review by mimowo — COMMENTED** [review]
LGTM, just small cleanups

File: pkg/controller/job/job_controller_test.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:1866]: this looks too cryptic:

```suggestion
				start := metav1.Time(fakeClock.Now().Sub(time.Second))
```

I think the original code was trying to truncate too, but I don't think that's important. But if you really want it, you can change the clock instead:

```golang
fakeClock := clocktesting.NewFakeClock(time.Now().Truncate(time.Second))
```

File: pkg/controller/job/job_controller.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller.go:1296]: /ack

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1883]: /Ack

File: pkg/controller/job/job_controller_test.go
Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1866]: `metav1.Time(fakeClock.Now().Sub(time.Second))` This can't be done right ? The `.Sub` returns a `time.Duration` where as the `metav1.NewTime` needs a `time.Time`

Instead can I just do `metav1.NewTime(clocktesting.NewFakeClock(time.Now().Truncate(time.Second)).Now())` instead ? @alculquicondor

Comment (alculquicondor) [reply pkg/controller/job/job_controller_test.go:1866]: oh, I guess it is `time.Add(-time.Second)`.

Comment (harshanarayana) [reply pkg/controller/job/job_controller_test.go:1866]: /Ack. Took care of both.

**Comment by mimowo** [comment]
LGTM

**Review by soltysh — APPROVED** [review]
/lgtm

**Pull request merged**
